{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, random\n",
    "from os import path, makedirs\n",
    "from urllib.error import HTTPError\n",
    "from time import sleep\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import Entrez, SeqIO\n",
    "from Bio.Alphabet import generic_dna\n",
    "\n",
    "Entrez.email = # your email here\n",
    "bed_file = \"http://jaspar.genereg.net/download/bed_files/MA0852.2.bed\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def expand_query(query: pd.Series) -> pd.DataFrame:\n",
    "    \"\"\" Parses query formatted for UCSC Biologic database \"\"\"\n",
    "    # expand query\n",
    "    parser = re.compile(r\"^([a-z]{2,4}\\d{,2})(?:\\_)(chr\\d{,2}[A-Z]{,3})(?:[^\\:]*):(\\d+)-(\\d+)(?:\\()([\\-\\+])(?:\\))\")\n",
    "    results = query.apply(parser.findall).apply(lambda r: list(sum(r, ()))).apply(pd.Series)\n",
    "    return results\n",
    "\n",
    "def load_jasper_bed_files(filename):\n",
    "    \"\"\" loads jasper bed files which contain\n",
    "        jasper motif-id and query of motif\n",
    "    \"\"\"\n",
    "    final_cols = [\"motif-id\", \"organism\", \"genome\", \"chromosome\", \"start\", \"stop\", \"strand\"]\n",
    "    # read csv\n",
    "    locus_file = pd.read_csv(filename,\n",
    "                             sep=\"\\t\",\n",
    "                             header=None)\n",
    "\n",
    "    # 3: query -> organism, chormosome, start, stop, strand\n",
    "    result_df = expand_query(locus_file.iloc[:,3])\n",
    "\n",
    "    # name columns\n",
    "    result_df.columns = final_cols[2:]\n",
    "    \n",
    "    # convert start and stop\n",
    "    result_df['start'] = result_df['start'].apply(int)\n",
    "    result_df['stop'] = result_df['stop'].apply(int)\n",
    "\n",
    "    # motif id is part of file name\n",
    "    result_df[\"motif-id\"] = filename[-12:-4]\n",
    "\n",
    "    # switch org names to Latin\n",
    "    result_df[\"organism\"] = \"homo sapiens\"\n",
    "\n",
    "    # reorder columns\n",
    "    return result_df[final_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_df = load_jasper_bed_files(bed_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif-id</th>\n",
       "      <th>organism</th>\n",
       "      <th>genome</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MA0852.2</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1304891</td>\n",
       "      <td>1304904</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MA0852.2</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2311718</td>\n",
       "      <td>2311731</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MA0852.2</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3403994</td>\n",
       "      <td>3404007</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MA0852.2</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3624782</td>\n",
       "      <td>3624795</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MA0852.2</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6601250</td>\n",
       "      <td>6601263</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   motif-id      organism genome chromosome    start     stop strand\n",
       "0  MA0852.2  homo sapiens   hg38       chr1  1304891  1304904      +\n",
       "1  MA0852.2  homo sapiens   hg38       chr1  2311718  2311731      -\n",
       "2  MA0852.2  homo sapiens   hg38       chr1  3403994  3404007      -\n",
       "3  MA0852.2  homo sapiens   hg38       chr1  3624782  3624795      +\n",
       "4  MA0852.2  homo sapiens   hg38       chr1  6601250  6601263      -"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def id_to_refseq(org, chrom):\n",
    "    id_switch = {\n",
    "            'homo sapiens':\n",
    "            lambda chrom:\n",
    "            f\"NC_0000{(int(chrom)):02d}\" if\n",
    "            chrom.isnumeric() and (0 < int(chrom) < 23) else\n",
    "            \"NC_000023\" if chrom == \"X\" else\n",
    "            \"NC_000024\" if chrom == \"Y\" else\n",
    "            None,\n",
    "    }\n",
    "    return id_switch[org](chrom)\n",
    "\n",
    "def fetch_chromosome(iden):\n",
    "    \"\"\" fetch fasta sequence from NCBI given id \"\"\"\n",
    "    try:\n",
    "        with Entrez.efetch(db='nucleotide', id=iden, rettype='fasta') as handle:\n",
    "            rec = SeqIO.read(handle, 'fasta')\n",
    "            rec.Alphabet = generic_dna\n",
    "            return rec\n",
    "    except HTTPError:\n",
    "        return None\n",
    "\n",
    "def download_needed_chromosomes(datf) -> int:\n",
    "    \"\"\"save chromosome sequences as fasta files in sequences folder\n",
    "    returns: number of sequences successfully saved\n",
    "    \"\"\"\n",
    "    saved = 0\n",
    "    # for each chromosome:\n",
    "    for org, chrom in datf[['organism', 'chromosome']] \\\n",
    "                           .groupby(['organism', 'chromosome']).count().index:\n",
    "\n",
    "        # get id for looking in NCBI\n",
    "        refseq_id = id_to_refseq(org, chrom.strip('chr'))\n",
    "\n",
    "        outfilename = \"tmp/\" + org.replace(\" \", \"-\") +\\\n",
    "                          chrom.strip('chr') + \".fasta\"\n",
    "            \n",
    "        if not path.isdir(\"./tmp/\"):\n",
    "            makedirs(\"./tmp/\")\n",
    "\n",
    "        # fetch sequence from NCBI\n",
    "        big_sequence = fetch_chromosome(refseq_id)\n",
    "\n",
    "        if not big_sequence:\n",
    "            print('error: in {} {} refseq_id {}'.format(org, chrom, refseq_id))\n",
    "\n",
    "        else:\n",
    "            # save sequence\n",
    "            SeqIO.write(big_sequence, outfilename, 'fasta')\n",
    "\n",
    "            # update status\n",
    "            saved += 1\n",
    "            print('saved', big_sequence.description)\n",
    "\n",
    "            # ensure requests do not go to NCBI too fast\n",
    "            sleep(1)\n",
    "    return saved\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saved NC_000001.11 Homo sapiens chromosome 1, GRCh38.p12 Primary Assembly\n",
      "saved NC_000010.11 Homo sapiens chromosome 10, GRCh38.p12 Primary Assembly\n",
      "saved NC_000011.10 Homo sapiens chromosome 11, GRCh38.p12 Primary Assembly\n",
      "saved NC_000012.12 Homo sapiens chromosome 12, GRCh38.p12 Primary Assembly\n",
      "saved NC_000013.11 Homo sapiens chromosome 13, GRCh38.p12 Primary Assembly\n",
      "saved NC_000014.9 Homo sapiens chromosome 14, GRCh38.p12 Primary Assembly\n",
      "saved NC_000015.10 Homo sapiens chromosome 15, GRCh38.p12 Primary Assembly\n",
      "saved NC_000016.10 Homo sapiens chromosome 16, GRCh38.p12 Primary Assembly\n",
      "saved NC_000017.11 Homo sapiens chromosome 17, GRCh38.p12 Primary Assembly\n",
      "saved NC_000018.10 Homo sapiens chromosome 18, GRCh38.p12 Primary Assembly\n",
      "saved NC_000019.10 Homo sapiens chromosome 19, GRCh38.p12 Primary Assembly\n",
      "saved NC_000002.12 Homo sapiens chromosome 2, GRCh38.p12 Primary Assembly\n",
      "saved NC_000020.11 Homo sapiens chromosome 20, GRCh38.p12 Primary Assembly\n",
      "saved NC_000021.9 Homo sapiens chromosome 21, GRCh38.p12 Primary Assembly\n",
      "saved NC_000022.11 Homo sapiens chromosome 22, GRCh38.p12 Primary Assembly\n",
      "saved NC_000003.12 Homo sapiens chromosome 3, GRCh38.p12 Primary Assembly\n",
      "saved NC_000004.12 Homo sapiens chromosome 4, GRCh38.p12 Primary Assembly\n",
      "saved NC_000005.10 Homo sapiens chromosome 5, GRCh38.p12 Primary Assembly\n",
      "saved NC_000006.12 Homo sapiens chromosome 6, GRCh38.p12 Primary Assembly\n",
      "saved NC_000007.14 Homo sapiens chromosome 7, GRCh38.p12 Primary Assembly\n",
      "saved NC_000008.11 Homo sapiens chromosome 8, GRCh38.p12 Primary Assembly\n",
      "saved NC_000009.12 Homo sapiens chromosome 9, GRCh38.p12 Primary Assembly\n",
      "saved NC_000023.11 Homo sapiens chromosome X, GRCh38.p12 Primary Assembly\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "23"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "download_needed_chromosomes(bed_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "LENGTH = 1000\n",
    "RAND_LIM = 500\n",
    "\n",
    "def choose_row(r, chrom):\n",
    "    if r[\"strand\"] == \"+\":\n",
    "        return str(chrom[r[\"rstart\"]-1:r[\"rstop\"]].seq)\n",
    "    else: \n",
    "        return str(chrom[r[\"rstart\"]-1:r[\"rstop\"]].seq.reverse_complement())\n",
    "\n",
    "def search_chromosome(chrom: SeqIO.SeqRecord, start: pd.Series,\n",
    "                      stop: pd.Series, strand: pd.Series) -> tuple:\n",
    "    \"\"\" returns (motif, mstart, mstop) \"\"\"\n",
    "    \n",
    "    # length of motif\n",
    "    len_motifs = (stop - start) + 1 # plus 1 because 0 vs. 1 indexing\n",
    "    \n",
    "    rstart = start - len_motifs.apply(lambda d: np.random.randint(0, LENGTH - d))\n",
    "    rstop = rstart + LENGTH\n",
    "    # get randomized indents within a set range\n",
    "\n",
    "    # select motif +/- indents\n",
    "    motifs = pd.concat([rstart, rstop, strand], keys=[\"rstart\", \"rstop\", 'strand'], axis=1)\n",
    "    motifs = motifs.apply(lambda r: choose_row(r, chrom), axis=1)\n",
    "\n",
    "    # return motif, start index from selected sequence, and\n",
    "    # stop index from selected sequence\n",
    "    return (motifs, start - rstart, start - rstart + len_motifs)\n",
    "\n",
    "def load_sequences(datf):\n",
    "    \"\"\" load chromosomes and return dataframe of sequences with motifs \"\"\"\n",
    "    dirname = \"tmp/\"\n",
    "    seqdat = pd.DataFrame()\n",
    "    for org, chrom in datf[['organism', 'chromosome']] \\\n",
    "                           .groupby(['organism', 'chromosome']).count().index:\n",
    "\n",
    "        chrom_file = dirname + org.replace(\" \", \"-\") +\\\n",
    "                     chrom.strip(\"chr\") + \".fasta\"\n",
    "        chrom_record = SeqIO.read(chrom_file, 'fasta')\n",
    "\n",
    "        # get rows for organism and chromosome\n",
    "        startstops = datf.loc[(datf['organism'] == org) & (datf['chromosome'] == chrom)]\n",
    "        # retrive motif + indent\n",
    "        motifs, mstarts, mstops = search_chromosome(chrom_record,\n",
    "                                                    startstops[\"start\"],\n",
    "                                                    startstops[\"stop\"],\n",
    "                                                    startstops[\"strand\"])\n",
    "        rows = pd.concat([startstops, motifs, mstarts, mstops], axis=1)\n",
    "        rows.columns = [\"motif-id\", \"organism\", \"genome\", \"chromosome\", \"start\",\n",
    "                       \"stop\", \"strand\", \"seq\", \"mstart\", \"mstop\"]\n",
    "        seqdat = seqdat.append(rows, ignore_index=True)\n",
    "\n",
    "    return seqdat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_df = load_sequences(bed_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif-id</th>\n",
       "      <th>organism</th>\n",
       "      <th>genome</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>seq</th>\n",
       "      <th>mstart</th>\n",
       "      <th>mstop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MA0852.2</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1304891</td>\n",
       "      <td>1304904</td>\n",
       "      <td>+</td>\n",
       "      <td>CCCTCACCCCACTGCACTACTGCAGCCCATCCAGGTCTGGCACCCA...</td>\n",
       "      <td>359</td>\n",
       "      <td>373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MA0852.2</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2311718</td>\n",
       "      <td>2311731</td>\n",
       "      <td>-</td>\n",
       "      <td>GTGGGGCACAGGGTGGGGCAGAAACTAGGGAGGTGGGGGAGCCCCC...</td>\n",
       "      <td>542</td>\n",
       "      <td>556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MA0852.2</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3403994</td>\n",
       "      <td>3404007</td>\n",
       "      <td>-</td>\n",
       "      <td>GGGGGTGTAGGGAGCACGGGGGCGGGCGCCCGTCTGCAGCCTCCTG...</td>\n",
       "      <td>24</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MA0852.2</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3624782</td>\n",
       "      <td>3624795</td>\n",
       "      <td>+</td>\n",
       "      <td>GCACACACATGTTCCACCTCGAACTGCGAGCGCTGACCGGGATCTC...</td>\n",
       "      <td>482</td>\n",
       "      <td>496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MA0852.2</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6601250</td>\n",
       "      <td>6601263</td>\n",
       "      <td>-</td>\n",
       "      <td>CCCGGAGCCTGTAAACATGGTCACATGTTAAGAGGTTTTCTTCCAG...</td>\n",
       "      <td>980</td>\n",
       "      <td>994</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   motif-id      organism genome chromosome    start     stop strand  \\\n",
       "0  MA0852.2  homo sapiens   hg38       chr1  1304891  1304904      +   \n",
       "1  MA0852.2  homo sapiens   hg38       chr1  2311718  2311731      -   \n",
       "2  MA0852.2  homo sapiens   hg38       chr1  3403994  3404007      -   \n",
       "3  MA0852.2  homo sapiens   hg38       chr1  3624782  3624795      +   \n",
       "4  MA0852.2  homo sapiens   hg38       chr1  6601250  6601263      -   \n",
       "\n",
       "                                                 seq  mstart  mstop  \n",
       "0  CCCTCACCCCACTGCACTACTGCAGCCCATCCAGGTCTGGCACCCA...     359    373  \n",
       "1  GTGGGGCACAGGGTGGGGCAGAAACTAGGGAGGTGGGGGAGCCCCC...     542    556  \n",
       "2  GGGGGTGTAGGGAGCACGGGGGCGGGCGCCCGTCTGCAGCCTCCTG...      24     38  \n",
       "3  GCACACACATGTTCCACCTCGAACTGCGAGCGCTGACCGGGATCTC...     482    496  \n",
       "4  CCCGGAGCCTGTAAACATGGTCACATGTTAAGAGGTTTTCTTCCAG...     980    994  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_df.to_csv('./MA0852_2/MA0852.2SEQS1k.csv', sep=\",\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
